src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java

```diff
@@ -0,0 +1,277 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static com.linecorp.armeria.server.VirtualHost.ensureHostnamePatternMatchesDefaultHostname;
+import static com.linecorp.armeria.server.VirtualHost.normalizeDefaultHostname;
+import static com.linecorp.armeria.server.VirtualHost.normalizeHostnamePattern;
+import static java.util.Objects.requireNonNull;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.InputStreamReader;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+import javax.net.ssl.SSLException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.NativeLibraries;
+
+import io.netty.handler.codec.http2.Http2SecurityUtil;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.ApplicationProtocolConfig.Protocol;
+import io.netty.handler.ssl.ApplicationProtocolConfig.SelectedListenerFailureBehavior;
+import io.netty.handler.ssl.ApplicationProtocolConfig.SelectorFailureBehavior;
+import io.netty.handler.ssl.ApplicationProtocolNames;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.ssl.SupportedCipherSuiteFilter;
+
+/**
+ * Contains information for the build of the virtual host.
+ * 
+ * @see ChainedVirtualHostBuilder
+ * @see VirtualHostBuilder
+ */
+@SuppressWarnings("rawtypes")
+abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractVirtualHostBuilder.class);
+    
+    protected static final ApplicationProtocolConfig HTTPS_ALPN_CFG = new ApplicationProtocolConfig(
+            Protocol.ALPN,
+            // NO_ADVERTISE is currently the only mode supported by both OpenSsl and JDK providers.
+            SelectorFailureBehavior.NO_ADVERTISE,
+            // ACCEPT is currently the only mode supported by both OpenSsl and JDK providers.
+            SelectedListenerFailureBehavior.ACCEPT,
+            ApplicationProtocolNames.HTTP_2,
+            ApplicationProtocolNames.HTTP_1_1);
+
+    protected static final String LOCAL_HOSTNAME;
+
+    static {
+        // Try the '/usr/bin/hostname' command first, which is more reliable.
+        Process process = null;
+        String hostname = null;
+        try {
+            process = Runtime.getRuntime().exec("hostname");
```

Comment:
That's actually what I wrote. :-)

Windows also has `hostname` command which does the same job, so I think it's fairly safe to do this.